test/unit/gateways/realex_test.rb

```diff
@@ -51,6 +51,41 @@ def setup
     @amount = 100
   end
 
+  def test_initialize_sets_refund_and_credit_hashes
+    refund_secret = 'refund'
+    rebate_secret = 'rebate'
+
+    options = {
+      login: @login,
+      password: @password,
+      rebate_secret: rebate_secret,
+      refund_secret: refund_secret
+    }
+
+    gateway = RealexGateway.new(options)
+
+    set_options = gateway.instance_variable_get(:@options)
+
+    assert set_options[:refund_hash] == Digest::SHA1.hexdigest(rebate_secret)
+    assert set_options[:credit_hash] == Digest::SHA1.hexdigest(refund_secret)
+  end
+
+  def test_initialize_with_nil_refund_and_rebate_secrets
+    options = {
+      login: @login,
+      password: @password,
+      rebate_secret: nil,
+      refund_secret: nil
+    }
+
+    gateway = RealexGateway.new(options)
+
+    set_options = gateway.instance_variable_get(:@options)
+
+    assert_false set_options.key?(:refund_hash)
+    assert_false set_options.key(:credit_hash)
```

Comment:
If the above gateway creation is made into an assert statement, then are these tests still needed? I'm not sure that they really get at the behavior that we're trying to test here.

Reply:
Why is it important for the behavior of this gateway that these hashes aren't added as keys to `options`? In my mind, this is an implementation detail rather than a description of the overall desired behavior.

Reply:
Hmm. I think we only want these hashes added if the corresponding secrets are provided. If the implementation of ```initialize``` were to add keys with some bad value when the secrets are not provided, the rest of the code would behave as if a ```refundhash``` is available when it's not.

Reply:
Maybe my question is more of a higher level one. Theoretically, my understanding is that tests should favor testing behavior over implementation. If I were to describe what this test is designed to check it would be "the gateway is successfully created even when `nil` is passed in as the values for the secrets." Based on that description, it seems to me that testing whether the `refund_hash` and `credit_hash` do not exist as keys in `@options` is more of an implementation detail that shouldn't necessarily be tested.

From an actual code perspective, it doesn't seem to be an issue at all if the hashes do get added with a `nil` value:
![image](https://user-images.githubusercontent.com/42787645/62487501-07782a00-b790-11e9-8755-252ee9815c73.png)

Reply:
I designed the test to test the behavior of the ```initialize``` method. I'd say that setting/not setting fields relates to behavior, rather than implementation. The behavior is "what" the code does, whereas the implementation is "how" the code does it. So, you wouldn't want to have your test check that your method is calling some private method, because that's an implementation detail. The current code does pass when passing in a ```nil``` secret, but the test is covering the broader result of calling ```initialize```. So going back to my previous example, if someone were to come in and refactor ```initialize``` in such a way that a bad value were added for ```options[:refund_hash]``` instead of not adding a key/value pair at all, the test would fail.

Reply:
Could you explain what you mean by "a bad value"? I had thought that you meant `nil`, but it sounds like you might mean something else.

Reply:
Sure. Any value calculated without a secret would be a "bad" value, because presumably there's no valid way to calculate a hash without a secret. So, hypothetically, if we switched to a different method for calculating the hash and it returned something (instead of erroring out like with the current method) when given ```nil```, we'd have a bogus hash, and the test would catch that unexpected behavior.

Reply:
I tend to agree that testing the observable behavior when initialized with bad input is useful here. Basically this test, with these assertions, is defining what the original intended behavior was: we only try to compute and store the hash values if the correct inputs are passed.

Reply:
Thanks for clarifying what behavior we're trying to test @bayprogrammer!